### PR TITLE
Add status messages for artifact installation

### PR DESCRIPTION
**Pull request by nhz2**
This PR replaces the artifact download progress bar with a status message if an installation step takes over half a second.

Here the unpacking step for Qt6Base takes over half a second.

This is required for #4001, where installing may take up to 60 seconds on some systems.

https://github.com/user-attachments/assets/5756c92f-0fca-40e2-9f4f-4e613647aa5b

### CODE REVIEW

File: src/Operations.jl [hunk 1]
```diff
@@ -835,13 +838,11 @@ function download_artifacts(ctx::Context;
...
     end
     push!(pkg_roots, dirname(env.project_file))
-    used_artifact_tomls = Set{String}()
-    download_jobs = Function[]
+    download_jobs = Dict{SHA1, Function}()
 
     print_lock = Base.ReentrantLock() # for non-fancyprint printing
 
-    # name -> (state, bar) where state is :ready, :running, :done, or :failed
-    download_states = Dict{String, DownloadState}()
```
Comment (nhz2) [inline src/Operations.jl:844]: Artifact names are not globally unique, so I changed the key to the artifact tree hash.

**Comment by nhz2** [comment]
This is what I get on my Windows laptop, where installing and unpacking are much slower.

https://github.com/user-attachments/assets/dc97e292-9886-4493-8590-d4cdaa0c25cc

File: src/Operations.jl [hunk 2]
```diff
@@ -852,47 +853,54 @@ function download_artifacts(ctx::Context;
...
+                    () -> begin
+                        try
+                            dstate.state = :running
+                            ret()
+                            if !fancyprint
+                                @lock print_lock printpkgstyle(io, :Downloaded, "artifact $rname $(MiniProgressBars.pkg_format_bytes(dstate.bar.max; sigdigits=1))")
                             end
+                        catch
+                            dstate.state = :failed
+                            rethrow()
```
Comment (IanButterworth) [inline src/Operations.jl:894]: Should we set progress status to `failed` here and keep failed jobs around in the list until all complete/failed?

**Review by IanButterworth — COMMENTED** [review]
Looks great

**Comment by KristofferC** [comment]
It's unclear to me what "installing" refers to here. The whole process of downloading, unpacking etc is installing to me. Could that be made more precise?

**Comment by KristofferC** [comment]
Maybe the "Downloading artifacts" should be changed to "Installing artifacts" to show that it encompasses the whole procedure.

File: src/Operations.jl (hunk 2 above)
Comment (nhz2) [reply src/Operations.jl:894]: I tried this but got strange formatting when many artifacts failed to install.

```
 Installing artifacts ╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0/83
  XML2                     failed
  Qt6Wayland               failed
  Gettext                  failed
  Xorg_xcb_util_wm         failed
  Libuuid                  failed
  EarCut                   failed
  Xorg_xcb_util            failed
  Vulkan_Loader            failed
 Installing artifacts ╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0/83
  XML2                     failed
  Qt6Wayland               failed
  Gettext                  failed
 Installing artifacts ╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0/83
 Installing artifacts ╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0/83
  XML2                     failed
 Installing artifacts ╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0/83
  XML2                     failed
 Installing artifacts ╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0/83
```

Comment (IanButterworth) [reply src/Operations.jl:894]: Just need to change the counting of rows to include to include :failed

Comment (nhz2) [reply src/Operations.jl:894]: Is it an issue if more rows than can fit in the terminal are printed?

**Comment by nhz2** [comment]
Here is what the new version looks like.


https://github.com/user-attachments/assets/22f3fda4-95df-44ee-bd95-d5d4d5f15d4d

File: src/Operations.jl [hunk 3]
```diff
@@ -852,65 +853,78 @@ function download_artifacts(ctx::Context;
...
-
         if fancyprint
             t_print = Threads.@spawn begin
                 try
                     print(io, ansi_disablecursor)
                     first = true
                     timer = Timer(0, interval=1/10)
                     # TODO: Implement as a new MiniMultiProgressBar
-                    main_bar = MiniProgressBar(; indent=1, header = "Downloading artifacts", color = :green, mode = :int, always_reprint=true)
+                    main_bar = MiniProgressBar(; indent=1, header = "Installing artifacts", color = :green, mode = :int, always_reprint=true)
```
Comment (IanButterworth) [inline src/Operations.jl:911]:
```suggestion
                    main_bar = MiniProgressBar(; indent=2, header = "Installing artifacts", color = :green, mode = :int, always_reprint=true)
```
Given `Installing` is one shorter. To align right

**Review by IanButterworth — APPROVED** [review]

**Pull request merged**